### PR TITLE
#18402 allow dotParse to parse dotAssets

### PR DESCRIPTION
**Pull request by wezell**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 1]
```diff
@@ -1,159 +1,239 @@
...
-      return null != asset.getFileAsset()? asset.getFileAsset().getAbsolutePath():null;
-    } catch (ResourceNotFoundException e) {
-        Logger.warn(this.getClass(), " - unable to resolve " + templatePath + " getting this: "+ e.getMessage() );
-        if(e.getStackTrace().length>0){
-          Logger.warn(this.getClass(), " - at " + e.getStackTrace()[0]);
+
+        final StringTokenizer tokens = new StringTokenizer(templatePath, StringPool.FORWARD_SLASH);
+
+        if (tokens.countTokens() < 2) {
+            String errorMessage = String.format("No resource found for [%s]", templatePath);
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:191]: Codacy Issue found: [Local variable 'errorMessage' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 2]
```diff
@@ -1,159 +1,239 @@
...
         }
-        throw e;
-    } catch (DotSecurityException  e) {
-        Logger.warn(this.getClass(), " - unable to resolve " + templatePath + " getting this: "+ e.getMessage() );
-        if(e.getStackTrace().length>0){
-            Logger.warn(this.getClass(), " - at " + e.getStackTrace()[0]);
+
+        tokens.nextToken();
+        final String inodeOrIdentifier = tokens.nextToken();
+        final String fieldVar = tokens.hasMoreTokens() ? tokens.nextToken() : DotAssetContentType.ASSET_FIELD_VAR;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:197]: Codacy Issue found: [Avoid declaring a variable if it is unreferenced before a possible exit point.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 3]
```diff
@@ -1,159 +1,239 @@
...
+        return Optional.of(Tuple.of(identifier, fieldVar));
+
+
     }
-  }
 
-    private void throwNotResourceNotFoundException(final RenderParams params, final String templatePath) {
-        final String errorMessage = String.format("Not found %s version of [%s]", params.mode, templatePath);
-        throw new ResourceNotFoundException(errorMessage);
+    private Optional<Tuple2<Identifier, String>> resolveFileAsset(final Context context, final RenderParams params,
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:221]: Codacy Issue found: [Avoid unused method parameters such as 'context'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 4]
```diff
@@ -1,159 +1,239 @@
...
-
-      return null != asset.getFileAsset()? asset.getFileAsset().getAbsolutePath():null;
-    } catch (ResourceNotFoundException e) {
-        Logger.warn(this.getClass(), " - unable to resolve " + templatePath + " getting this: "+ e.getMessage() );
-        if(e.getStackTrace().length>0){
-          Logger.warn(this.getClass(), " - at " + e.getStackTrace()[0]);
+
+        final StringTokenizer tokens = new StringTokenizer(templatePath, StringPool.FORWARD_SLASH);
+
+        if (tokens.countTokens() < 2) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:190]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 5]
```diff
@@ -1,159 +1,239 @@
...
+                        ? ConversionUtils.toLong(request.getAttribute(contentLanguageId), defaultLanguageId)
+                        : defaultLanguageId;
+    }
+
+
+    @Override
+    String resolveTemplatePath(final Context context, final Writer writer, final RenderParams params, final String[] arguments) {
+
+
+        String templatePath = arguments[0];
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:80]: Codacy Issue found: [Local variable 'templatePath' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 6]
```diff
@@ -1,159 +1,239 @@
...
+        if (!shortOpt.isPresent()) {
+            return Optional.empty();
         }
-        throw new ResourceNotFoundException(e);
-    } catch (Exception e) {
-        Logger.warn(this.getClass(), " - unable to resolve " + templatePath + " getting this: "+ e.getMessage() );
-        if(e.getStackTrace().length>0){
-            Logger.warn(this.getClass(), " - at " + e.getStackTrace()[0]);
+
+        ShortyId shorty = shortOpt.get();
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:204]: Codacy Issue found: [Local variable 'shorty' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 7]
```diff
@@ -1,159 +1,239 @@
...
+                                        params.language.getId(), params.user, false)
+                        : Optional.ofNullable(
+                                        APILocator.getContentletAPI().find(shorty.longId, params.user, params.mode.respectAnonPerms));
+
+        if (!conOpt.isPresent()) {
+            return Optional.empty();
         }
-        throw new DotStateException(e);
+        
+        Identifier identifier = APILocator.getIdentifierAPI().find(conOpt.get().getIdentifier());
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:215]: Codacy Issue found: [Local variable 'identifier' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 8]
```diff
@@ -1,159 +1,239 @@
...
+
+
     }
-  }
 
-    private void throwNotResourceNotFoundException(final RenderParams params, final String templatePath) {
-        final String errorMessage = String.format("Not found %s version of [%s]", params.mode, templatePath);
-        throw new ResourceNotFoundException(errorMessage);
+    private Optional<Tuple2<Identifier, String>> resolveFileAsset(final Context context, final RenderParams params,
+                    String templatePath) throws DotDataException, DotSecurityException {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:222]: Codacy Issue found: [Avoid reassigning parameters such as 'templatePath'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 9]
```diff
@@ -1,159 +1,239 @@
...
 
-      ContentletVersionInfo contentletVersionInfo = APILocator.getVersionableAPI()
-              .getContentletVersionInfo(identifier.getId(), languageId);
 
-      if (contentletVersionInfo == null  ||  UtilMethods.isNotSet(contentletVersionInfo.getIdentifier())  || contentletVersionInfo.isDeleted()) {
 
-        final long defaultLang = APILocator.getLanguageAPI().getDefaultLanguage().getId();
-        if (defaultLang != languageId) {
-          contentletVersionInfo = APILocator.getVersionableAPI().getContentletVersionInfo(identifier.getId(), defaultLang);
+    private Optional<Tuple2<Identifier, String>> resolveDotAsset(final Context context, final RenderParams params,
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:180]: Codacy Issue found: [Avoid unused method parameters such as 'context'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 10]
```diff
@@ -1,159 +1,239 @@
...
+                optIdAndField = resolveFileAsset(context, params, templatePath);
+            }
+            
+            
+            if (!optIdAndField.isPresent()) {
+                throwNotResourceNotFoundException(params, templatePath);
+            }
+            
+            
+            Tuple2<Identifier, String> idAndField = optIdAndField.get();
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:99]: Codacy Issue found: [Local variable 'idAndField' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java [hunk 11]
```diff
@@ -1,159 +1,239 @@
...
+        // if we have a host
+        if (templatePath.startsWith(hostIndicator)) {
+            templatePath = templatePath.substring(hostIndicator.length(), templatePath.length());
+            String hostName = templatePath.substring(0, templatePath.indexOf('/'));
+            templatePath = templatePath.substring(templatePath.indexOf('/'), templatePath.length());
+            host = APILocator.getHostAPI().resolveHostName(hostName, user, params.mode.respectAnonPerms);
+        }
+
+
+        Identifier identifier = APILocator.getIdentifierAPI().find(host, templatePath);
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/directive/DotParse.java:234]: Codacy Issue found: [Local variable 'identifier' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5627236)

**Comment by dsilvam** [comment]
Any type of test for this change?

**Review by freddyucv — COMMENTED** [review]
any test?

**Review by jdotcms — APPROVED** [review]

**Review by freddyucv — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

**Pull request merged**